### ISSUE TITLE
High Variance on RNN-based PPO agents

### ISSUE THREAD

**Issue by fiorenza2** [ray-project/ray#4030]
### System information
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Ubuntu 16.04
- **Ray installed from (source or binary)**: Source
- **Ray version**: 0.6.2
- **Python version**: 3.6.7
- **Exact command to reproduce**: ```use_lstm: True```

### Describe the problem

I've noticed that during PPO training, policy loss variance is incredibly high when an RNN is added compared to when it is left out. See below, where the first picture is from a custom scenario I'm training (red is normal, green is a custom GRU, and pink is LSTM) , and the second is from stateless CartPole (orange is with LSTM, blue is without, NB: I forced the examples script to not load a LSTM).

<img src="https://user-images.githubusercontent.com/15997485/52674906-f3f4b080-2f1c-11e9-9dca-67004c84a8cf.png" alt="drawing" height="200"/>

<img src="https://user-images.githubusercontent.com/15997485/52674886-df181d00-2f1c-11e9-95a3-0673d026e13c.png" alt="drawing" height="200"/>

In the case of CartPole, the policy with LSTM still learns how to 'solve' the scenario of course, but I've noticed that performance doesn't seem as high as expected in more complex scenarios.

Therefore I was wondering if this higher variance on the policy loss was **expected behaviour** when adding an RNN network, and if so why this is the case?

BTW, I tried to do a bit of a deep dive, and noticed that the high variance mainly stems from the advantage calculation. I therefore added an RNN to the value function as well (similar approach to [here](https://github.com/ray-project/ray/issues/2666)) but this did not curb this behaviour.

**Comment by ericl** [comment]
Could you elaborate on what you mean by the variance coming from the advantage calculation? Is the value function much worse at predicting than compared with use_lstm: False? That would be unexpected.

**Comment by ericl** [comment]
Actually, that makes sense. With layer sharing, the VF prediction error improves a lot. However, it did not seem to actually reduce the variance of the policy loss that much.

Have you tried increasing the training batch size?

**Comment by fiorenza2** [comment]
So I noticed the value function could explain more variance when adding the RNN to it, but this did not affect the variance on the policy loss, nor did it improve actual agent performance.

In terms of the variance analysis I did, I noted that the policy loss term comprises basically the advantages multiplied by some logp_ratio (the latter of which can be clipped). The logp_ratio is stable, and slowly decays over time, but taking the mean over the advantages gives the same variance observed in the policy loss, which suggests this is the source. I had a quick look at ```compute_advantages``` function in the ```postprocessing.py``` file but nothing jumped out at me immediately.

I will play around with the batch size and let you know if that improves things.

**Comment by fiorenza2** [comment]
Had a go increasing batch sizes, but if anything the problem seems to worsen surprisingly 😮 . Here's the main hyperparams I'm using (basically a variant of the PPO paper's Atari hyperparameters ([Table 4](https://arxiv.org/pdf/1707.06347.pdf))):
```
        num_workers: 23
        num_gpus: 1
        lambda: 0.95
        lr: 0.0001
        sgd_minibatch_size: 2944    # or 736, since 23 workers * 128 = 2944, 23 workers * 32 = 736
        observation_filter: MeanStdFilter
        horizon: 128
        gamma: 0.99
        clip_param: 0.1
        vf_loss_coeff: 1.0
        entropy_coeff: 0.01
        num_sgd_iter: 3
```

Here's a plot comparing, with batch size of 736 in grey, and 2944 in blue:
<img src="https://user-images.githubusercontent.com/15997485/52715557-a1ef7180-2f94-11e9-919c-7444307b48b5.png" alt="drawing" height="200"/>

**Comment by ericl** [comment]
Huh, that's puzzling, it's hard to imagine why value target prediction
would improve but not decrease the variance of the advantages.

Re: hyperparameters, I think you need to actually increase train_batch_size
as well, a batch of 4k by default with a minibatch of 2900 seems quite
small.

**Comment by ericl** [comment]
Well, another thing I noticed is that the policy is actually more unstable, i.e. the min and max rewards higher variance than without the LSTM. So that may be in part responsible for the high value variance.

**Comment by fiorenza2** [comment]
So in my scenario I noticed that the variance around the learnt policy (max/min rewards) to be relatively similar bewteen with/without LSTM. Also maybe I've got the wrong end of the stick, but wouldn't one expect the value function loss to also increase/have higher variance if the learnt policy itself is less stable, since the agent will be doing 'less predictable' actions in the environment?

Good point about base-lining, I believe there might be something I can look into, and see if the higher variance exists there too. 

Also maybe this is completely wrong since I need to get more familiar with the code/Tensorflow, but is it somehow possible that during the ```compute_advantages``` function certain correspondences between a given state's predicted score (i.e., Q(s,a)) and baseline value (i.e., V(s)) (where Advantage = Q(s,a) - V(s)) are being mixed up because of the padding?

**Comment by ericl** [comment]
It's entirely possible there is some bug in sequencing (though I haven't been able to find it after some cursory audits). The way postprocessing works is that it runs on rollout fragments generated by policy evaluation. This is done quite early in the process, before padding is introduced. So it's probably not likely that advantages get mis-aligned due to padding.

I also tried out some different hyperparameters (such as using "simple_optimizer": True which is less likely to have a bug since it does much less of the batch slicing needed for multi-gpu) and the variance can change quite a lot (though it's not clear this affected learning performance).

**Comment by ericl** [comment]
@fiorenza2 , did you ever figure out if the variance is expected for the env?

**Comment by ericl** [comment]
Closing this since it can't be reproduced -- feel free to reopen.

**Comment by fiorenza2** [comment]
Hi Eric,

So did a bit more experimentation, and I now believe that increase in variance I observed was in-keeping with the environment. 

I was able to achieve similar variance in non-RNN policy networks by putting certain constraints on the agent (i.e., forcing it to explore new states). This suggests that the RNNs were having a similar effect, but I would need to view the rollouts to fully confirm.
Furthermore I did NOT notice a stacking effect (i.e. LSTM + constraints giving double the variance); instead the variance seemed rather dichotomous, either the small or large plots I showed above, nothing in between.

So whilst the variance issue seems relatively moot now, RNNs were still decreasing overall performance, but this will need to merit further investigation. If I have any updates about this I will let you know.
